README
$Id: README,v 1.2 2007/01/20 07:28:38 comdog Exp $

You can install this using in the usual Perl fashion

	perl Makefile.PL
	make
	make test
	make install
	
The documentation is in the module file.  Once you install
the file, you can read it with perldoc.

	perldoc Mac::iTerm::LaunchPad
	
The `new-iterm` script installs whereever you told the Makefile.PL
to install programs. Without a prefix that defauts to the value you
set up when you installed Perl. That is probably something like 
/usr/bin or /usr/bin/perl.

If you want to read the source before you install it, you can use
perldoc directly on the module file.

	perldoc  lib/LaunchPad.pm
	
This module is also in CVS on SourceForge

	http://sourceforge.net/projects/brian-d-foy/


Enjoy, 

brian d foy, bdfoy@cpan.org

Changes
1.01 - Sat Jan 20 00:58:16 2007
	* First release of script to create new iTerm sessions in specified
	directories, but without a lot of typing

1.07 - Fri Mar 16 18:16:41 2007
	* Upgraded Mac::Glue prereq to 1.28 (after the Intel fixes)

1.06 - Thu Jan 25 14:27:10 2007
	* turn on strictures to make CPANTS happy

1.05 - Sat Jan 20 16:55:35 2007
	* Added some bits to the Makefile.PL to create the iTerm glue if
	it isn't already there. If it cannot finally load the iTerm glue,
	it dies, although it did create the Makefile. The tests will
	fail in that case (at least until you manually install the iTerm
	glue). See Mac::Glue for details about using gluemac.
	* This resolves RT # 24473 from stennie. Thanks :)

1.04 - Sat Jan 20 01:29:20 2007
	* Added more to the docs to explain the new-iterm program
	* Fixed up some things that made it into the pre-CVS-imported dir
	I was accidently working in (two demerits for that!).

1.02 - Sat Jan 20 01:16:33 2007
	* Doc patch to fill in the DESCRIPTION portion



lib/LaunchPad.pm
#!/usr/bin/perl
use strict;

package Mac::iTerm::LaunchPad;

our $VERSION = 1.007;

#sprintf "%d.%03d", q$Revision: 2201 $ =~ m/(\d+) \. (\d+)/x;

=head1 NAME

new-iterm - open a new iTerm window with one or more tabs

=head1 SYNOPSIS

	---Frontmost finder directory, defaulting to desktop
	% new-iterm

	---Frontmost finder directory, using special alias
	% new-iterm finder
	
	---Named directory
	% new-iterm /Users/brian/Dev

	---Named directory with
	% new-iterm ~/Dev
	
	---Multiple tabs 
	% new-iterm ~/Dev ~/Foo ~/Bar
	
	---Aliases, predefined in code
	% new-iterm music applications
	
	---Aliases, defined in your own ~/.new-iterm-aliases
	% new-iterm foo bar baz
	
	---Any combination
	% new-iterm finder /Users/brian/Dev music ~/Pictures foo
	
	---From other commands
	% perldoc -l Mac::Glue | xargs dirname | xargs new-iterm
	
=head1 DESCRIPTION

This script opens a new iTerm window and creates a tab for each
directory. Within each tab, it changes to that directory. It allows
you to use aliases (not the filesystem sort, the nickname sort) to
save on typing. Without arguments it finds the frontmost finder window
and uses that as the directory. The special directory named "finder"
does the same thing (so you're stuck if you have a directory with that
name: give it a different alias).

=head2 Modulino

This script is actually a modulino. It's a symlink to the
C<Mac::iTerm::LaunchPad> module. That module figures out if it's run
as a script or included as a module and does the right thing. If you
want to change the program, edit the module.

If you don't like that idea, use the included F<scripts/new-iterm>
program which does the same thing without the symlink.

=head2 Aliases

You can define aliases in the F<~/.new-iterm-aliases> file. The file
is line-oriented and has the alias followed by its directory. You can
use the ~ home directory shortcut. 

	#alias	directory
	cpan /mirrors/MINICPAN
	dev	~/Dev
	paypal ~/Personal/Finances/PayPal
	
Since Mac OS X uses a case insenstive (though preserving) file system,
case doesn't matter. If you tricked Mac OS X into using something else,
use the right case and remove the C<lc()> in the code.

=head3 Default aliases

=over 4

=item desktop - the Desktop folder of the current user ( ~/Desktop )

=item home - home directory of the current user ( ~ )

=item music - music directory of the current user ( ~/Music )

=item applications - music directory of the current user ( ~/Applications )

=item finder - the directory of the frontmost finder window (defaults to Desktop)

=back

=head1 TO DO

=over 4

=item switch to choose session name (currently just default)

=item switch to specify tabs or new windows?

=back

=head1 AUTHOR

brian d foy, C<< <bdfoy@cpan.org> >>

Inspired by a script from Chris Nandor (http://use.perl.org/~pudge/journal/32199) 
which was inspired by a script from Curtis "Ovid" Poe 
(http://use.perl.org/~Ovid/journal/32086).

=head1 COPYRIGHT AND LICENSE

Copyright 2007, brian d foy.

You may use this program under the same terms as Perl itself.

Some parts come from Chris Nandor and are noted in the source. They are
available under the same license.

=head1 SEE ALSO

iTerm - http://iterm.sourceforge.net/

=cut

# defaults
my %Aliases = qw(
	desktop  		~/Desktop
	home     		~
	music    		~/Music
	applications	/Applications
	);

_run() unless caller;

sub _run {

_init();
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #	
# argument processing

foreach my $arg ( @ARGV )
	{
	my $cwd = do {
		if( defined $arg )
			{
			# don't lc() if you have a case sensitive file system
			$arg = lc( exists $Aliases{$arg} ? $Aliases{$arg} : $arg );
			$arg = _get_finder_dir() if $arg eq 'finder';
			
			if( -d $arg ) { $arg }
			else          { die "$arg isn't a directory!\n" }
			}
		else
			{
			_get_finder_dir();
			}
		};
	
	_launch_iterm( $cwd );
	}
}

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #	
sub _init
	{
	@ARGV = ( undef ) unless @ARGV; # finder window special case;

	foreach my $key ( keys %Aliases ) { $Aliases{$key} =~ s/^~/$ENV{HOME}/ }
	
	print Dumper ( \%Aliases );
	
	if( open my($fh), "<", "$ENV{HOME}/.new-iterm-aliases" )
		{
		while( <$fh> )
			{
			s/^\s|\s$//g;
			s|(?<!\\)#|.*|;
			next unless /\S\s+\S/;
			my( $alias, $dir ) = map { lc } split;
			$dir =~ s/^~/$ENV{HOME}/;
			$Aliases{$alias} = $dir;
			}
		}	
	}
	
sub _get_finder_dir
	{
	use Mac::Files;

	# from Chris Nandor
	my $finder = new Mac::Glue 'Finder';
	my $cwd = $finder->prop(target => window => 1)->get(as => 'alias');
	$cwd ||= FindFolder(kUserDomain, kDesktopFolderType); # default to Desktop
	$cwd =~ s/'/'\\''/g;
	$cwd;
	}
	
BEGIN {
use Mac::Glue ':all';

my $iterm = eval { Mac::Glue->new( 'iTerm' ) };
if( $@ ) { die "Could not load iTerm definitions for Mac::Glue!: $@" }

$iterm->activate;
my $term = $iterm->make( new => 'terminal' );
my $session = 1;

sub _launch_iterm
	{
	my $cwd = shift;
	
	$term->Launch( session => 'default' );
	$term->obj( session => $session++ )->write(text => "cd '$cwd'");
	}
}

Makefile.PL
# $Id: Makefile.PL 2201 2007-03-17 01:16:15Z comdog $
use ExtUtils::MakeMaker;

eval "use Test::Manifest 1.14";

sub MY::postamble 
	{
	<<'MAKE';
install :: new-iterm-install
	
new-iterm-install:
	$(ECHO) "Installing new-iterm symlink to module"
	/bin/ln -sfv $(INSTALLSITELIB)/Mac/iTerm/LaunchPad.pm $(INSTALLBIN)/new-iterm 
	$(CHMOD) 755 $(INSTALLSITELIB)/Mac/iTerm/LaunchPad.pm
MAKE
	}
	
WriteMakefile(
	'NAME'	       => 'Mac::iTerm::LaunchPad',
	'VERSION_FROM' => 'lib/LaunchPad.pm',
	'LICENSE'	   => 'perl',
	'AUTHOR'       => 'brian d foy <bdfoy@cpan.org>',
	
	'PREREQ_PM'    => { 
		'Test::More'     => '0',
		'Mac::Files'     => '0',
		'Mac::Glue'      => '1.28',
		'Mac::Processes' => '0',
		},

	'PM'           => {
		'lib/LaunchPad.pm'         => '$(INST_LIBDIR)/LaunchPad.pm',
		},

	'MAN3PODS'     => {
		'lib/LaunchPad.pm' => '$(INST_MAN3DIR)/Mac::iTerm::LaunchPad.3',
		},

	clean  => { FILES    => q|Mac-iTerm-LaunchPad-*| },

	);

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# An auto installer should already have installed Mac::Glue
{
require Mac::Glue;

my $iterm = eval { Mac::Glue->new( 'iTerm' ) };

if( $@ ) 
	{
	require Mac::Processes; Mac::Processes->import;
	
	my $iterm_path = LSFindApplicationForInfo( undef, undef, 'iTerm.app' );

	if( $iterm_path ) 
		{
		print <<"HERE";
To create the Mac::Glue bindings for iTerm you need to run:

	sudo gluemac $iterm_path

HERE

		print "Would you like to run this now? (y/N) ";

		chomp( my $confirm = <> );

		if( lc( $confirm ) eq 'y' ) 
			{
			system 'sudo', 'gluemac', $iterm_path;
			$iterm = eval { Mac::Glue->new( 'iTerm' ) };
			}
		}
	else
		{
		die "Could not find iTerm. Ensure you have it and try again\n";
		}
	}

unless( defined $iterm )
	{
	die "Cannot continue: Ensure that the iTerm glue for Mac::Glue is installed\n";
	}

}

1;